README.md
# VSCode-AzureAutomationExtension
This extension will make you capable of working in Azure Automation without leaving Visual Studio Code.

## Before use
Before use, please follow the guide in the link below, to create an application in your Azure AD, to be used as access for this extension.
https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-create-service-principal-portal

Make sure the application is added to your Automation Account as 'Contributor'

NEW: Make sure to update all the settings under 'Azure Automation configuration' by copying to your custom settings.

## Runbook view container
In the side bar, a new view has been added to give an overview of your runbooks.

When clicking on a runbook it will automatically fetch your draft version and create as a file locally in you open folder.
However, if you have a published version available, it will show as a child - so expand the runbook item and click 'Published' to get you published version of the runbook, to edit.
Below is a quick view of how the tree structure is supposed to look for your runbooks.
 - 'runbookname'
   - 'Published' (only if a published version exists)
   - 'Tags' (Shows a view of current tags associated with your runbok)

## Functionality
A list of the functionality available in this extension

 - Create Runbook (Using a template or from scratch)
 - Open existing runbook
 - Save Draft
 - Publish
 - Publish and Run
 - Run
 - Create new Azure variable asset
 - Create new Azure credential asset
 - Insert Azure variable asset
 - Insert Azure credential asset

### Create Runbook
Types supported: PowerShell and Python2
Create a new runbook in your defined Azure Automation account.
Support for a template is possible. Under 'User Settings' just input the name of your template runbook, and any new runbook created through this extension, will be prefilled with the same content as that template runbook.
If you leave the name blank, a blank runbook will be created.

### Open existing runbook
It is now posible to open a runbook already in your Automation Account, and start working on it inside VS Code.

### Save draft
Save a draft copy to the automation account

### Publish
Publish the runbook, will save a draft of the current editor content and then publish

### Publish and Run
Will do the same as the above, but afterwards it will execute the runbook for you to test functionality

### Run
Will execute the published version of runbook open in editor

### Create new Azure variable / credential asset
Create a new variable or credential asset.
Based on the 'DualVars' setting, this will either create 1 or 2 assets. Insterting code into your runbook automatically

![New asset](https://github.com/Maaaarden/VSCode-AzureAutomationExtension/raw/master/resources/newasset.gif)

### Insert Azure variable / credential asset
Insert an existing variable or credential asset.
Based on the 'DualVars' setting, this will either use 1 or 2 assets. Insterting code into your runbook automatically

## Job output
The extension makes use of output consoles in VS Code. When using it to execute a runbook in Azure (or any of your hybridworker groups), all outputs are shown inside VS Code.

Error output is automatically highlighted if any. 

## Background
This extension was made with specific use inside my organisation for eye. A few of the functions have been designed specifically for this.
Since version 0.6.4, it has been posible to turn off these functionalities. Trying to make this a generic usable as posible.

## Feedback
Feedback on this functionality is more than welcome - as is changes to this, to support a broader usecase.

I would really appreciate if you could bring usecases for the following examples, and requests to how it should work:

 - Parameters
    - It is posible to pull the parameters from the API, and thus be able to build some kind of functionality around that to support entering parameters when trying to run a runbook
 - Draft / Published
    - I would like to hear about your usage of these 2 versions of the runbook. And to hear if you have any wishes to support that way of working.

CHANGELOG.md
# Change Log
All notable changes to the "azureautomation" extension will be documented in this file.

## [1.0.14] - 2021-12-15
### Fixed
 - Fixed the output channel for PS7 runbooks.

## [1.0.11] - 2021-11-25
### Removed
 - Removed the option to manualle override API version. This has caused some API call to fail.
   All API are updated with latest version as per MS Documentation.

## [1.0.8] - 2021-11-10
### Fixed
 - Fixed a hardcoded region for runbooks, causing issues for automation accounts not in westeurope.

### Added
 - Added support for runbook runtime versions. Support for both PowerShell and Pyhton.

## [1.0.0] - 2019-02-20
### Fixed
 - Fixed an error that surfaced, when writing to disk. Causing "open runbook" not to work.

## [0.9.0] - 2018-10-29
### Added
 - Added support for Python2 runbooks. Create, Save Draft, Publish and run.
   Variable/Credential insert commands not supported.

## [0.8.0] - 2018-09-10
### Fixed
 - Fixed an issue with the character '+' in clientSecret, causing user to not get an oAuth token.
 
## [0.7.2] - 2018-09-06
### Added
 - Added functionality to check for settings before each command, to ensure all relevant settings have been filled
 
### Fixed
 - Fixed an issue causing 'Open runbook' from view, not to work

## [0.7.1] - 2018-09-05
### Fixed
 - Fixed an issue causing nobody to be able to get oAuth tokens for login.

## [0.7.0] - 2018-07-25
### Added
 - Added a new view container to the side bar with a view of runbooks in your Automation Account
 - Added a 'Refresh' button to the view
 - Added a 'Create new runbook' button to the view

### Fixed
 - Fixed an issue when having more than 100 runbooks, the rest would not be shown in 'Open runbook from Azure' command

## [0.6.7] - 2018-06-13
### Fixed
 - Fixed and error still preventing job execution when no hybridworkers present

## [0.6.6] - 2018-06-12
### Added
 - Added 'Open runbook from azure' functionality

### Fixed
 - Fixed problem when executing runbook on Azure, with no HybridWorkers on automation account

## [0.6.5] - 2018-05-23
### Changed
 - Added support for 'Integer' and 'Boolean' variable types

## [0.6.4] - 2018-05-21
### Changed
 - Added configuration option for 'DualVars' notation
    Change the value of 'dualVars' under configuration to 'false', to make the extension only make use of 1 variable/credential
 - Added code to support for both dual and single var notation

## [0.6.2] - 2017-09-20
### Fixed
 - Fixed a bug, where creation of variables / credentials didn't work.

## [0.6.1] - 2017-08-31
### Changed
 - Removed usage of 'azureconfig.json' file with settings for connection to Azure.
 - Added use of settings in VS Code instead

### Removed
 - Removed use of name and mail properties to be able to create runbooks.

## [0.5.2] - 2017-08-31
### Added
 - Added info about job output to readme


modules/AzureAutomation.js

/**
 * This function gets the bearer token needed to communicate to Azure API
 * @param   {Function}  next  Callback function
 * @return  {Object}          Returns an object with the token.
 */
var getOauthToken = function (next) {
  var vscode = require('vscode')
  var request = require('request')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")

  //var clientSecret = _.replace(azureconfig.clientSecret, new RegExp('\\+','g'),'%2B')
  var clientSecret = azureconfig.clientSecret
  request({
    url: `https://login.microsoftonline.com/${azureconfig.tenantId}/oauth2/token?api-version=1.0`,
    body: `grant_type=client_credentials&resource=https%3A%2F%2Fmanagement.core.windows.net%2F&client_id=${encodeURIComponent(azureconfig.clientId)}&client_secret=${encodeURIComponent(clientSecret)}`,
    headers: {
      'content-type': 'application/x-www-form-urlencoded'
    }
  }, function (error, response, body) {
    if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
      console.log(body)
      return vscode.window.showErrorMessage('Could not get OAuth Token! Error:' + error)
    }
    var bodyParsed = JSON.parse(body)
    var token = bodyParsed.access_token
    return next({value: 'Bearer ' + token})
  })
}

var getAutomationAccountRegion = function () {
  return new Promise((resolve) => {
    var vscode = require('vscode')
    var request = require('request')
    var azureconfig = vscode.workspace.getConfiguration("azureautomation")
    var LogEngine = require('./LogEngine.js')

    getOauthToken(function (token) {
      request.get({
        url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}?api-version=2015-10-31`,
        headers: {
          'Authorization': token.value,
          'content-type': 'application/json'
        }
      }, function (error, response, body) {
        if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
          LogEngine.writeLog('getAutomationAccountRegion', response)
          return vscode.window.showErrorMessage(`Could not get Automation Account info for ${azureconfig.automationAccount}`)
        }

        var bodyParsed = JSON.parse(body)
        resolve(bodyParsed.location)
      })
    })
  })
}

var getRunbookInfo = function (runbookName) {
  return new Promise((resolve) => {
    var vscode = require('vscode')
    var request = require('request')
    var azureconfig = vscode.workspace.getConfiguration("azureautomation")
    var LogEngine = require('./LogEngine.js')
    
    getOauthToken(function (token) {
      request.get({
        url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks/${runbookName}?api-version=2015-10-31`,
        headers: {
          'Authorization': token.value,
          'content-type': 'application/json'
        }
      }, function (error, response, body) {
        if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
          LogEngine.writeLog('getRunbookInfo', response)
          return vscode.window.showErrorMessage('Could not get runbook info for ' + runbookName + '! Error: ' + error.message)
        }
        var bodyParsed = JSON.parse(body)

        resolve(bodyParsed)
      })
    })
  })
}

/**
 * Gets a list of all runbooks in Azure and returns an array containing
 * only the names without the .ps1 extension.
 * @param   {Function}  next    Callback function
 * @return  {Array}             Returns an array with runbook names.
 */
var getListOfRunbooks = function (next, skip = false, runbookNames = false) {
  var vscode = require('vscode')
  var request = require('request')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")
  var _ = require('lodash')
  var LogEngine = require('./LogEngine.js')

  if(!skip) {
    skip = ''
  }
  if(!runbookNames) {
    runbookNames = []
  }

  getOauthToken(function (token) {
    request.get({
      url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks?api-version=2015-10-31${skip}`,
      headers: {
        'Authorization': token.value,
        'content-type': 'application/json'
      }
    }, function (error, response, body) {
      var bodyParsed = JSON.parse(body)
      if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
        LogEngine.writeLog('getListOfRunbooks', response)
        return vscode.window.showErrorMessage('Could not get list of Runbook from Azure Automation! Error: ' + bodyParsed.error.message)
      }
      var i = 0
      _.forEach(bodyParsed.value, function (runbookObject) {
        runbookNames.push(runbookObject.name)
        i++
      })
      if(bodyParsed.nextLink && i == bodyParsed.value.length) {
        skip = '&' + bodyParsed.nextLink.split('&').pop()
        getListOfRunbooks(next, skip, runbookNames)
      } else {
        return next(runbookNames)
      }
    })
  })
}

/**
 * Saves the focused runbook tab as a draft in Azure. It returns
 * an object wether the runbook exists or not. {success: true/false}
 * @param   {Function}  next    Callback function
 * @return  {Object}            Returns an object with property 'exists' true/false
 */
var saveAsDraft = function (next) {
  var vscode = require('vscode')
  var request = require('request')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")
  var _ = require('lodash')
  var document = vscode.window.activeTextEditor.document
  var LogEngine = require('./LogEngine.js')

  if (document.isUntitled) {
    return vscode.window.showWarningMessage('Please save your runbook locally before saving to Azure.')
  }
  // First splits the string on '\', second takes the last value in the array, finally replaces '.ps1' with nothing (removes it).
  var rbPath = _.last(_.split(document.fileName, '\\'))
  var runbookName = rbPath.substr(0, rbPath.lastIndexOf('.'))
  /**if(runbookType == 'PowerShell') {
    var runbookName = _.replace(_.last(_.split(document.fileName, '\\')), '.ps1', '')
  } else if(runbookType == 'Python2') {
    var runbookName = _.replace(_.last(_.split(document.fileName, '\\')), '.py', '')
  }*/
  var fileText = document.getText()

  getOauthToken(function (token) {
    request.put({
      url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks/${runbookName}/draft/content?api-version=2015-10-31`,
      headers: {
        'Authorization': token.value,
        'content-type': 'application/json'
      },
      body: fileText
    }, function (error, response, body) {
      
      if (response.statusCode === 202) {
        vscode.window.setStatusBarMessage('Draft saved successfully.', 3100)
        LogEngine.writeLog('saveAsDraft', response)
        // vscode.window.showInformationMessage('Draft saved successfully.')
        return next({ success: true })
      } else if (response.statusCode === 404) {
        return next({ success: false, reason: 'Runbook does not exist in Azure' })
      } else if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
        LogEngine.writeLog('saveAsDraft', response)
        console.log(body)
        return vscode.window.showErrorMessage('An error accoured while trying to save the draft in Azure Cloud.')
      } else {
        return vscode.window.showErrorMessage('An error accoured while trying to save the draft in Azure Cloud.')
      }
    })
  })
}

/**
 * Creates a new runbook in Azure unpublished.
 * @param   {Function}  next    Callback function
 */
var createAzureRunbook = function (runbookType, runbookRuntime, next) {
  var vscode = require('vscode')
  var request = require('request')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")
  var _ = require('lodash')
  var document = vscode.window.activeTextEditor.document
  var rbPath = _.last(_.split(document.fileName, '\\'))
  var runbookName = rbPath.substr(0, rbPath.lastIndexOf('.'))
  var LogEngine = require('./LogEngine.js')

  if(runbookType == 'PowerShell') {
    if(runbookRuntime.replace(' (preview)', '') == '7.1')
      runbookType = 'PowerShell7'
  } else {
    if(runbookRuntime.replace(' (preview)', '') == '3.8.0') {
      runbookType = 'Python3'
    } else {
      runbookType = 'Python2'
    }
  }

  if (document.isUntitled) {
    return vscode.window.showErrorMessage('Please save your runbook locally before saving to Azure.')
  }
  doesRunbookExist(runbookName, function (runbookExists) {
    if (runbookExists) {
      vscode.window.showErrorMessage('Runbook name already exists.')
      return next()
    } else {
      getAutomationAccountRegion()
      .then(region => {
        getOauthToken(function (token) {
          var requestData = {
              'tags': {
                'CreatedFrom': 'VSCode Extension'
              },
              'properties': {
                'runbookType': runbookType,
                'draft': {
                  'inEdit': false
                }
              },
              'location': region
            }
          request.put({
            url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks/${runbookName}?api-version=2015-10-31`,
            headers: {
              'Authorization': token.value,
              'Content-Type': 'application/json'
            },
            json: requestData
          }, function (error, response, body) {
            if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
              LogEngine.writeLog('createAzureRunbook', response)
              return vscode.window.showErrorMessage('An error occured while trying to create the runbook in Azure Cloud. Error: ' + error.message)
            }
            if (response.statusCode === 201) {
              vscode.window.setStatusBarMessage('Runbook created in Azure Cloud.', 3100)
              //vscode.window.showInformationMessage('Runbook created in Azure Cloud.')
              return next()
            } else {
              LogEngine.writeLog('createAzureRunbook', response)
              return vscode.window.showErrorMessage('Could not create Runbook in Azure Cloud..')
            }
          })
        })
      })
    }
  })
}

/**
 * This function creates a new local runbook. It downloads and adds the BS
 * custom Azure Runbook content to the newly created file.
 * @param   {String}    runbookName The name of the runbook you wish to create.
 * @param   {Function}  next        Callback function
 */
var createLocalRunbook = function (runbookName, runbookType, existing=false, published=false, next) {
  var vscode = require('vscode')
  var request = require('request')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")
  var fs = require('fs')
  var LogEngine = require('./LogEngine.js')

  var hasWorkspace = !!vscode.workspace.rootPath
  if (!hasWorkspace) {
    return vscode.window.showErrorMessage('No workspace found. Please open a folder before creating a new Runbook.')
  }

  let publishedV = published == false ? 'draft/' : ''

  if (existing) {
    getOauthToken(function (token) {
      request.get({
        url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks/${runbookName}/${publishedV}content?api-version=2015-10-31`,
        headers: {
          'Authorization': token.value
        }
      }, function (error, response, body) {
        if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
          LogEngine.writeLog('createLocalRunbook', response)
          return vscode.window.showErrorMessage('Could not get runbook from Azure Cloud. Error: ' + error.message)
        }
        if(runbookType == 'PowerShell' || runbookType == 'PowerShell7') {
          var path = vscode.workspace.rootPath + `\\${runbookName}.ps1`
        } else if(runbookType == 'Python' || runbookType == 'Python3') {
          var path = vscode.workspace.rootPath + `\\${runbookName}.py`
        }
        
        fs.writeFile(path, body, function() {
          vscode.workspace.openTextDocument(path).then(doc => {
            vscode.window.showTextDocument(doc)
            setTimeout(function () {
              next()
            }, 2000)
          })
        })
      })
    })
  } else if (azureconfig.templateName != "") {
    getOauthToken(function (token) {
      request.get({
        url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks/${azureconfig.templateName}/content?api-version=${azureconfig.apiVersion}`,
        headers: {
          'Authorization': token.value
        }
      }, function (error, response, body) {
        if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
          LogEngine.writeLog('createLocalRunbook', response)
          return vscode.window.showErrorMessage('Could not get template from Azure Cloud.')
        }
        if(runbookType == 'PowerShell' || runbookType == 'PowerShell7') {
          var path = vscode.workspace.rootPath + `\\${runbookName}.ps1`
        } else if(runbookType == 'Python' || runbookType == 'Python3') {
          var path = vscode.workspace.rootPath + `\\${runbookName}.py`
        }

        fs.writeFile(path, body, function() {
          vscode.workspace.openTextDocument(path).then(doc => {
            vscode.window.showTextDocument(doc)
            setTimeout(function () {
              next()
            }, 2000)
          })
        })
      })
    })
  } else {
    if(runbookType == 'PowerShell' || runbookType == 'PowerShell7') {
      var path = vscode.workspace.rootPath + `\\${runbookName}.ps1`
    } else if(runbookType == 'Python' || runbookType == 'Python3') {
      var path = vscode.workspace.rootPath + `\\${runbookName}.py`
    }

    fs.writeFile(path, "", function() {
      vscode.workspace.openTextDocument(path).then(doc => {
        vscode.window.showTextDocument(doc)
        setTimeout(function () {
          next()
        }, 2000)
      })
    })
  }
}

var publishRunbook = function (next) {
  var vscode = require('vscode')
  var request = require('request')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")
  var _ = require('lodash')
  var document = vscode.window.activeTextEditor.document
  var rbPath = _.last(_.split(document.fileName, '\\'))
  var runbookName = rbPath.substr(0, rbPath.lastIndexOf('.'))
  var LogEngine = require('./LogEngine.js')

  getOauthToken(function (token) {
    request.post({
      url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/runbooks/${runbookName}/draft/publish?api-version=2015-10-31`,
      headers: {
        'Authorization': token.value
      }
    }, function (error, response, body) {
      if (error || response.statusCode != 200 && response.statusCode != 201 && response.statusCode != 202) {
        LogEngine.writeLog('publishRunbook', response)
        return vscode.window.showErrorMessage('An error accoured while trying to publish the runbook in Azure Cloud.')
      }
      if (response.statusCode === 202) {
        vscode.window.setStatusBarMessage('Runbook successfully published.', 3100)
        vscode.commands.executeCommand('azureautomation.updateRunbookProvider')
        return next()
      } else {
        LogEngine.writeLog('publishRunbook', response)
        return vscode.window.showErrorMessage('Could not publish the runbook.')
      }
    })
  })
}

/**
 *
 * @param {*} token
 * @param {*} next
 */
var startPublishedRunbook = function (token, next) {
  var vscode = require('vscode')
  var request = require('request')
  var jobs = require('./AzureAutomationJobs.js')
  var azureconfig = vscode.workspace.getConfiguration("azureautomation")
  var guid = createGuid()
  var _ = require('lodash')
  var LogEngine = require('./LogEngine.js')

  var rbPath = _.last(_.split(vscode.window.activeTextEditor.document.fileName, '\\'))
  var runbookName = rbPath.substr(0, rbPath.lastIndexOf('.'))
  jobs.getHybridWorkerGroups(token, function (hybridWorkers) {
    if (!hybridWorkers) {
      request.put({
        url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/jobs/${guid}?api-version=2017-05-15-preview`,
        headers: {
          'Authorization': token
        },
        json: {
          'properties': {
            'runbook': {
              'name': '' + runbookName + ''
            },
            'parameters': { // Parameters of current open runbook.
            }
          }
        }
      }, function (error, response) {
        if (response.statusCode === 404 || error) {
          LogEngine.writeLog('startPublishedRunbook', response)
          vscode.window.showErrorMessage('Something went wrong, when trying to start the job.')
        }
        if (response.statusCode === 201) {
          return next(guid)
        }
        if (response.statusCode === 200) {
          vscode.window.showInformationMessage('Job already running. Usually this means a mishap in the code.')
        }
      })
    } else {
      vscode.window.showQuickPick(hybridWorkers)
    // eslint-disable-next-line no-undef
    .then(val => runOn = val)
    .then(function (runOn) {
      if(runOn.detail) {
        request.put({
          url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/jobs/${guid}?api-version=${azureconfig.apiVersion}`,
          headers: {
            'Authorization': token
          },
          json: {
            'properties': {
              'runbook': {
                'name': '' + runbookName + ''
              },
              'parameters': { // Parameters of current open runbook.
              },
              'runOn': runOn.detail
            }
          }
        }, function (error, response) {
          if (response.statusCode === 404 || error) {
            LogEngine.writeLog('startPublishedRunbook', response)
            vscode.window.showErrorMessage('Something went wrong, when trying to start the job.')
          }
          if (response.statusCode === 201) {
            return next(guid)
          }
          if (response.statusCode === 200) {
            vscode.window.showInformationMessage('Job already running. Usually this means a mishap in the code.')
          }
        })
      } else {
        request.put({
          url: `https://management.azure.com/subscriptions/${azureconfig.subscriptionId}/resourceGroups/${azureconfig.resourceGroup}/providers/Microsoft.Automation/automationAccounts/${azureconfig.automationAccount}/jobs/${guid}?api-version=${azureconfig.apiVersion}`,
          headers: {
            'Authorization': token
          },
          json: {
            'properties': {
              'runbook': {
                'name': '' + runbookName + ''
              },
              'parameters': { // Parameters of current open runbook.
              }
            }
          }
        }, function (error, response) {
          if (response.statusCode === 404 || error) {
            LogEngine.writeLog('startPublishedRunbook', response)
            vscode.window.showErrorMessage('Something went wrong, when trying to start the job.')
          }
          if (response.statusCode === 201) {
            return next(guid)
          }
          if (response.statusCode === 200) {
            vscode.window.showInformationMessage('Job already running. Usually this means a mishap in the code.')
          }
        })
      }
    })
    }
  })
}

var doesRunbookExist = function (runbookName, next) {
  var _ = require('lodash')
  getListOfRunbooks(function (runbookList) {
    if (_.includes(runbookList, runbookName)) {
      return next(true)
    } else {
      return next(false)
    }
  })
}

function createGuid () {
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
    var r = Math.random() * 16 | 0, v = c === 'x' ? r : (r & 0x3 | 0x8)
    return v.toString(16)
  })
}

module.exports = {
  getOauthToken: getOauthToken,
  getListOfRunbooks: getListOfRunbooks,
  saveAsDraft: saveAsDraft,
  createAzureRunbook: createAzureRunbook,
  createLocalRunbook: createLocalRunbook,
  publishRunbook: publishRunbook,
  doesRunbookExist: doesRunbookExist,
  startPublishedRunbook: startPublishedRunbook,
  getRunbookInfo: getRunbookInfo
}
